---
title: "MAGMA Celltyping tutorial: Using MAGMA to find causative celltypes for genetically complex traits"
author: "P Timshel"
date: "`r Sys.Date()`"
---

# Introduction

This package takes GWAS summary statistics & Single Cell Transcriptome specificity data (in EWCE's CTD format) as input. As output it calculates the associations between the GWAS trait and the celltypes. If you use this package then you must cite our paper in which this was described:

"Genetic Identification Of Brain Cell Types Underlying Schizophrenia", Nature Genetics, March 2018, Nathan G. Skene, Julien Bryois, Trygve E. Bakken, Gerome Breen, James J Crowley, Héléna A Gaspar, Paola Giusti-Rodriguez, Rebecca D Hodge, Jeremy A. Miller, Ana Muñoz-Manchado, Michael C O’Donovan, Michael J Owen, Antonio F Pardiñas, Jesper Ryge, James T R Walters, Sten Linnarsson, Ed S. Lein, Major Depressive Disorder Working Group of the Psychiatric Genomics Consortium, Patrick F Sullivan, Jens Hjerling-Leffler"

# Initialization
```{r}
wd <- "/raid5/projects/timshel/sc-genetics/sc-genetics/src/nb-magma_celltyping"
setwd(wd)
```


# Setup MAGMA

```{r}
### Install packages
# devtools::install_github(repo = 'NathanSkene/MAGMA_Celltyping') # main package
# devtools::install_github(repo = 'NathanSkene/EWCE')
library(MAGMA.Celltyping) # loads EWCE package

# library(R.utils) # maybe needed
```


## Set parameters to be used for the analysis

```{r}
# Set path the 1000 genomes reference data.
# genome_ref_dir = "/tools/magma/1.06/" # this does 
# if(!file.exists(sprintf("%s/g1000_eur.bed",genome_ref_dir))){
#     download.file("https://ctg.cncr.nl/software/MAGMA/ref_data/g1000_eur.zip",destfile=sprintf("%s.zip",genome_ref_dir))
#     unzip(sprintf("%s.zip",genome_ref_dir),exdir=genome_ref_dir)
# }
```

```{r}
genome_ref_path = "/tools/magma/1.06/g1000_eur/g1000_eur" # should include the 'prefix' plink files
```


# Prepare cell type specificity data

## Install and load all the required R packages and data

The EWCE package comes with a celltype specificity dataset which we use as an example.  The One2One package is used to obtain 1:1 orthologs.

```{r }
# Load the celltype data
data(ctd) # loaded from the EWCT package

# str(ctd) 
# list[<data_set>] --> list of 3
# list[<data_set>]["mean_exp"] # data.frame. genes x cell_types
# list[<data_set>]["specificity"] # matrix genes x cell_types
# list[<data_set>]["annot"] # character. cell labels
# x <- ctd[[1]][["mean_exp"]]
# x <- ctd[[1]][["specificity"]]
# x <- ctd[[1]][["annot"]]
```

```{r}
# Load the mouse to human 1:1 orthologs
data(ortholog_data_Mouse_Human)  # loaded from the EWCT package

# str(ortholog_data_Mouse_Human)
# List of 10
#  $ orthologs_all                    :'data.frame':	17328 obs. of  3 variables:
#   ..$ HomoloGene ID: chr [1:17328] "100" "1000" "10001" "10003" ...
#   ..$ human.symbol : chr [1:17328] "ETFA" "CLDN4" "HAUS2" "EFHC1" ...
#   ..$ mouse.symbol : chr [1:17328] "Etfa" "Cldn4" "Haus2" "Efhc1" ...
#  $ orthologs_one2one                :'data.frame':	16470 obs. of  3 variables:
#   ..$ HomoloGene ID: chr [1:16470] "100" "1000" "10001" "10003" ...
#   ..$ human.symbol : chr [1:16470] "ETFA" "CLDN4" "HAUS2" "EFHC1" ...
#   ..$ mouse.symbol : chr [1:16470] "Etfa" "Cldn4" "Haus2" "Efhc1" ...
#  $ species1_allGenes                : chr [1:18708] "3" "5" "6" "7" ...
#  $ species2_allGenes                : chr [1:18995] "3" "5" "6" "7" ...
#  $ shared_genes                     : chr [1:16754] "3" "5" "6" "7" ...
#  $ species1_present_species2_deleted: chr [1:1954] "254" "388" "776" "803" ...
#  $ species2_present_species1_deleted: chr [1:2241] "431" "826" "862" "1532" ...
#  $ species1_onceOnly_species2_dup   : chr [1:214] "10356" "104130" "104838" "10506" ...
#  $ species2_onceOnly_species1_dup   : chr [1:60] "100509" "110586" "11434" "11573" ...
# $ species2_dup_species1_dup        : chr [1:10] "78045" "86332" "129517" "41961" ...
```


## Prepare quantile groups for celltypes

First we need to calculate the quantile groups for each celltype within the single cell dataset. This is done using the `prepare.quantile.groups` function. If your single cell dataset is not from mouse, then change the specificity_species argument. If you wish to use a smaller number of bins then 

```{r }
ctd = prepare.quantile.groups(ctd,specificity_species="mouse",numberOfBins=40)
```

```{r}
# Examine how the quantile groups look
print(ctd[[1]]$quantiles[c("Gfap","Dlg4","Aif1"),])
print(table(ctd[[1]]$quantiles[,1]))

```


# Run MAGMA

## Download summary statistics file & check it is properly formatted

We need to have a summary statistics file to analyse as input. As an example download summary statistics for Fluid Intelligence, based on the UK Biobank, generated by Ben Neale's group.

The function `check.sumstats.formatted.for.magma` does some basic processing to get it into the right format.


```{r}
# gwas_sumstats_path <- "/raid5/projects/timshel/sc-genetics/sc-genetics/data/gwas_sumstats_raw/BMI_HEIGHT_Yengo2018/Meta-analysis_Locke_et_al+UKBiobank_2018.txt"
# ERROR - reading p-value file: non-numeric or non-integer value for sample size variable N on line 447477
# 	line: 6	6176144	rs12210959	T	C	0.7384	-0.0014	0.0022	 5.1e-01	6e+05

gwas_sumstats_path <- "/raid5/projects/timshel/sc-genetics/sc-genetics/data/gwas_sumstats_raw/BMI_HEIGHT_Yengo2018/bmi_yengo2018_magma_fmt.txt"

# gwas_sumstats_path <- "/projects/tp/tmp-bmi-brain/data/gwas/bmi_height_yengo_2018/Meta-analysis_Locke_et_al+UKBiobank_2018.txt.gz.tab"
```

```{r}
# gwas_sumstats_path = "20016.assoc.tsv"
# ### Download and unzip the summary statistics file
# if(!file.exists(gwas_sumstats_path)){
#     download.file("https://www.dropbox.com/s/shsiq0brkax886j/20016.assoc.tsv.gz?raw=1",destfile=sprintf("%s.gz",gwas_sumstats_path))
#     gunzip(sprintf("%s.gz",gwas_sumstats_path),gwas_sumstats_path)
# }
```



```{r }
# Format it (i.e. column headers etc)
# source("magma_celltyping_modified_functions.R")
# col_headers = format_sumstats_for_magma.PT(gwas_sumstats_path)
```



## Map SNPs to Genes

```{r, eval=TRUE}
source("magma_celltyping_modified_functions.R")
genesOutPath = map.snps.to.genes.PT(gwas_sumstats_path,genome_ref_path=genome_ref_path)
```

## Run the main cell type association analysis

```{r }
ctAssocs = calculate_celltype_associations(ctd,gwas_sumstats_path,genome_ref_path=genome_ref_path)
```

```{r}
plot_celltype_associations(ctAssocs, savePDF=F)
# plot_celltype_associations(ctAssocs, savePDF=T)
```








## Run the conditional cell type association analysis

```{r }
# Conditional analysis
ctCondAssocs = calculate_conditional_celltype_associations(ctd,gwas_sumstats_path,genome_ref_path=genome_ref_path)
```


```{r}
plot_celltype_associations(ctCondAssocs, savePDF=F)
```



# PT EXTRA DOCS

```{r}
?check_inputs_to_magma_celltype_analysis
```

