---
### YAML section (use # for comments)
title: "Sc-genetics: RolyPoly"
author: "Pascal N. Timshel, Pers Lab"
date: "`r Sys.time()`" # Sys.Date()
output:
  html_notebook: 
    df_print: paged
    number_sections: yes
    toc: yes
    toc_depth: 3
    toc_float: yes
  html_document:
    df_print: paged
    number_sections: yes
    toc: yes
    toc_depth: 3
    toc_float: yes
---


```{r}
### Extract the code from the notebook
# knitr::purl("nb-rolypoly_functions.Rmd", output = "rolypoly_functions.R", documentation = 1)
```


# Setup

```{r, message = FALSE, results='hide', warning=FALSE}
library(rolypoly)
library(tidyverse)
library(ggplot2)
```

```{r}
dir.sc_genetics.data <- "/projects/timshel/sc-genetics/sc-genetics/data" # no trailing slash
dir.ldfiles <- file.path(dir.sc_genetics.data, "rolypoly/EUR_LD_FILTERED_NONAN_R") # Linkage disequilibrium files

dir.sc_genetics_lib <- "/projects/timshel/sc-genetics/sc-genetics/src/lib/"
source(sprintf("%s/load_functions.R", dir.sc_genetics_lib)) # load sc-genetics library
```


# Notebook specific parameters
```{r}
wd <- "/projects/timshel/sc-genetics/sc-genetics/src/"
setwd(wd)
```

# GWAS
```{r}
file.gwas <- file.path(dir.sc_genetics.data, "gwas_sumstats/body_BMI_Locke2015.gwassumstats.rolypoly_fmt.tab.gz") # BMI
# file.gwas <- file.path(dir.sc_genetics.data, "gwas_sumstats/mental_SCZ_Ripke2014.gwassumstats.rolypoly_fmt.tab.gz") # SCZ
```

```{r}
df.gwas.rp <- read_gwas(file.gwas, exlcude.HLA=T, do.log_odds=F)
```




# Gene annotation

```{r}
# file.gene_annot <- file.path(dir.sc_genetics.data, "gene_annotations/gene_annotation.hsapiens_mmusculus_unique_orthologs.GRCh37.ens_v91.txt.gz") # all 1-1 human-mouse ortholog
file.gene_annot <- file.path(dir.sc_genetics.data, "gene_annotations/gene_annotation.hsapiens_all_genes.GRCh37.ens_v91.txt.gz") # all human genes
```

```{r}
df.gene_annot <- read_gene_annotation(file.gene_annot, delim="\t")
```

```{r}
df.gene_annot %>% head()
```


# Expression data

REQUIRED COLUMNS IN DATA FRAME: none (just cell-type/tissue names); rownames is genes

```{r}
### t-test data
# file.expr <- file.path(dir.sc_genetics.data, "expression/arc_lira/arc_lira.celltype_expr.ttest.hsapiens_orthologs.csv.gz") # arc_lira
# file.expr <- file.path(dir.sc_genetics.data, "expression/maca/maca.per_tissue.celltype_expr.ttest.hsapiens_orthologs.csv.gz") # maca per_tissue


### average expression
# file.expr <- file.path(dir.sc_genetics.data, "expression/maca/maca.per_tissue.celltype_expr.avg_expr.hsapiens_orthologs.csv.gz") # maca per_tissue
file.expr <- file.path(dir.sc_genetics.data, "expression/maca/maca.per_tissue_celltype.celltype_expr.avg_expr.hsapiens_orthologs.csv.gz") # maca per_tissue_celltype
```



```{r}
df.expr <- read_expression_data(file.expr=file.expr,
                                 method_specific_expr="average", 
                                 pos_transformation="square", 
                                 genes.filter=df.gene_annot$ensembl_gene_id,  
                                 col_genes="gene", 
                                 delim=","
                                 ) 
```

```{r}
as_tibble(df.expr) %>% head()
```


# Block annotation: window size, etc

```{r}
# REQUIRED COLUMNS IN DATA FRAME: c('chrom', 'start', 'end', 'label')

WINDOW_SIZE <- 10e3 # 10kb

df.block_annotation <- df.gene_annot %>% transmute(
  label=ensembl_gene_id,
  chrom=chromosome_name,
  start=start_position-WINDOW_SIZE, # TSS-WINDOW_SIZE
  end=start_position+WINDOW_SIZE # TSS+WINDOW_SIZE
)

df.block_annotation %>% head
```


# Rolling rolypoly

```{r}
df.gwas.rp.test <- df.gwas.rp %>% slice(1:1000)
```


```{r}
library(doParallel)
registerDoParallel(40) # half of the number of cores 
getDoParWorkers()
```



```{r, results='hide'}
# message = FALSE

# TODO: run with "highres" data

rm(rp)
system.time(rp <- rolypoly_roll(
  gwas_data = as.data.frame(df.gwas.rp), # df.gwas.rp.test
  block_annotation = as.data.frame(df.block_annotation),
  block_data = as.data.frame (df.expr.filtered.rp),
  ld_folder = dir.ldfiles,
  bootstrap_iters = 200, # use at least 200
  gwas_link_parallel = T,
  bootstrap_parallel = T
))
```


```{r}
# rp.scz.maca.tissue <- rp
# save(rp.scz.maca.tissue, file="out.rolypoly_objs.SCZ.RData")

# rp.arc_lira <- rp
# rp.maca.tissue <- rp

# save(rp.arc_lira, rp.maca.tissue, file="out.rolypoly_objs.BMI.RData")
# load(file="out.rolypoly_objs.BMI.RData")

# save.image(file="out.rolypoly_main-NB.RData")
# load(file="out.rolypoly_main-SCRIPT.RData")
```


## Results

Plot the $-log10(p)$ to rank tissues by the strength of their association
```{r, fig.show='asis', fig.width=7, fig.height=8}
p.pval_ranking <- plot_rolypoly_annotation_ranking(rp)
p.pval_ranking
```


Plot gamma estimate with 95\% confidence intervals
gamma: influence of cell-type-specific gene expression on the variance of GWAS effect sizes
```{r, fig.show='asis', fig.width=7, fig.height=8}
p.gamma_est <- plot_rolypoly_annotation_estimates(rp)
p.gamma_est
```

```{r}
rp$full_results$parameters %>% sort
```

Data frame with results from the bootstrap runs that calculates standard errors for these paramter estimates, p-values, and 95\% confidence intervals.
```{r}
rp$bootstrap_results %>% arrange(-bt_value) %>% head
```



```{r}
########################################################################################################################
############################################### Parallelization #########################################################
########################################################################################################################
```


# Parallelization: Inference on other expression data set

Without any parallelization rolypoly could take a couple hours to run. Much of the computation is spent linking SNPs to genes and reading in LD information. If one provides the a rolypoly object to the rolypoly parameter of the `rolypoly_roll` function call and a new object of block data, then only inference is performed. For example with the previous rolypoly object we would run inference on a new expression data set with the following call (assuming that the gene labels did not change between datasets).

```{r}
# OBS: assuming that the gene labels did not change between datasets.
# rp.new <- rolypoly_roll(
#   # some new set of expression data
#   rolypoly = rp,
#   block_data = new_sim_expression_data_normalized
# )
```

Thus, one can run precomputation (slow) once and then rerun inference (faster) on various expression matrices more quickly.




```{r}
########################################################################################################################
################################################## LEFTOVER ############################################################
########################################################################################################################
```


Check that all genes are found in the gene annotation file
Ortherwise we would have to filter out genes not in the gene annotation file...


















